### PR TITLE
Conform to API route signature

### PR DESCRIPTION
**Pull request by matthewp**
## Changes

- Implements the API Route Change proposal RFC: https://github.com/withastro/rfcs/pull/178
- Retains backwards compatibility, with warnings.
- Provides an `APIContext` to dynamic routes: `{ params, request }` as outlined in the RFC.

## Testing

- Tests added to ensure we retain backwards compatibility on the old signatures.
- Tests added for the new signature.

### CODE REVIEW

**Comment by sarah11918** [comment]
@matthewp This all looks great when it's time to merge!

My only question is, looking at the page itself, do you think visually the SSR stuff is separated/stands out enough from the non-SSR stuff? I'll leave that to be your call. Just wanted to make sure that it's also functioning for people who might be scanning the docs.

**Review by tony-sull — APPROVED** [review]
lgtm!  Great test coverage here, just added a couple notes to keep eslint happy

File: packages/astro/src/@types/astro.ts [hunk 1]
```diff
@@ -855,12 +855,27 @@ export interface AstroAdapter {
...
+	params: Params;
+	request: Request;
+}
+
 export interface EndpointOutput<Output extends Body = Body> {
 	body: Output;
 }
 
+interface APIRoute {
+	(context: APIContext, request: Request): EndpointOutput | Response;
```
Comment (bholmesdev) [inline packages/astro/src/@types/astro.ts:868]:
```suggestion
	(context: APIContext): EndpointOutput | Response;
```

Is there a reason `request` is passed as a second argument here? I assumed `APIContext` covered that...

File: packages/astro/test/fixtures/ssr-dynamic/src/pages/api/products/[id].js [hunk 2]
```diff
@@ -1,5 +1,5 @@
 
-export function get(params) {
+export function get({params}) {
```
Comment (bholmesdev) [inline packages/astro/test/fixtures/ssr-dynamic/src/pages/api/products/[id].js:2]:
```suggestion
export function get({ params }) {
```

nit: formatting

**Review by bholmesdev — COMMENTED** [review]
Just needs a changeset and a couple nits. Otherwise, our (re)quest for the perfect API is complete

**Pull request merged**